File: src/featurize/src/main/scala/Featurize.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark

import org.apache.spark.annotation.DeveloperApi
import org.apache.spark.ml.param._
import org.apache.spark.ml.util._
import org.apache.spark.ml.{Estimator, Pipeline, PipelineModel}
import org.apache.spark.sql._
import org.apache.spark.sql.types._

private object FeaturizeUtilities
{
  // 2^18 features by default
  val numFeaturesDefault = 262144
  // 2^12 features for tree-based or NN-based learners
  val numFeaturesTreeOrNNBased = 4096
}

object Featurize extends DefaultParamsReadable[Featurize]

/** Featurizes a dataset. Converts the specified columns to feature columns. */
class Featurize(override val uid: String) extends Estimator[PipelineModel] with MMLParams {

  def this() = this(Identifiable.randomUID("Featurize"))

  /** Feature columns - the columns to be featurized
    * @group param
    */
  val featureColumns: MapArrayParam = new MapArrayParam(this, "featureColumns", "Feature columns")

  /** @group getParam */
  final def getFeatureColumns: Map[String, Seq[String]] = $(featureColumns)

  /** @group setParam */
  def setFeatureColumns(value: Map[String, Seq[String]]): this.type = set(featureColumns, value)

  /** One hot encode categorical columns when true; default is true
    * @group param
    */
  val oneHotEncodeCategoricals: Param[Boolean] = BooleanParam(this,
    "oneHotEncodeCategoricals",
    "One-hot encode categoricals",
    true)

  /** @group getParam */
  final def getOneHotEncodeCategoricals: Boolean = $(oneHotEncodeCategoricals)

  /** @group setParam */
  def setOneHotEncodeCategoricals(value: Boolean): this.type = set(oneHotEncodeCategoricals, value)

  /** Number of features to hash string columns to
    * @group param
    */
  val numberOfFeatures: IntParam =
    IntParam(this,
      "numberOfFeatures",
      "Number of features to hash string columns to",
      FeaturizeUtilities.numFeaturesDefault)

  /** @group getParam */
  final def getNumberOfFeatures: Int = $(numberOfFeatures)

  /** @group setParam */
  def setNumberOfFeatures(value: Int): this.type = set(numberOfFeatures, value)

  /** Specifies whether to allow featurization of images */
  val allowImages: Param[Boolean] = BooleanParam(this, "allowImages", "Allow featurization of images", false)

  /** @group getParam */
  final def getAllowImages: Boolean = $(allowImages)

  /** @group setParam */
  def setAllowImages(value: Boolean): this.type = set(allowImages, value)

  /** Featurizes the dataset.
    *
    * @param dataset The input dataset to train.
    * @return The featurized model.
    */
  override def fit(dataset: Dataset[_]): PipelineModel = {
    val pipeline = assembleFeaturesEstimators(getFeatureColumns)
    pipeline.fit(dataset)
  }

  private def assembleFeaturesEstimators(featureColumns: Map[String, Seq[String]]): Pipeline = {
    val assembleFeaturesEstimators = featureColumns.map(newColToFeatures => {
      new AssembleFeatures()
        .setColumnsToFeaturize(newColToFeatures._2.toArray)
        .setFeaturesCol(newColToFeatures._1)
        .setNumberOfFeatures(getNumberOfFeatures)
        .setOneHotEncodeCategoricals(getOneHotEncodeCategoricals)
        .setAllowImages(getAllowImages)
    }).toArray

    new Pipeline().setStages(assembleFeaturesEstimators)
  }

  override def copy(extra: ParamMap): Estimator[PipelineModel] = {
    new Featurize()
  }

  @DeveloperApi
  override def transformSchema(schema: StructType): StructType =
    assembleFeaturesEstimators(getFeatureColumns).transformSchema(schema)

}

File: src/src/main/resources/mmlspark/Featurize.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class Featurize(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEstimator):
    """
    Featurizes a dataset.  Converts the specified columns to feature columns.

    Args:

        allowImages (bool): Allow featurization of images (default: false)
        featureColumns (dict): Feature columns
        numberOfFeatures (int): Number of features to hash string columns to (default: 262144)
        oneHotEncodeCategoricals (bool): One-hot encode categoricals (default: true)
    """

    @keyword_only
    def __init__(self, allowImages=False, featureColumns=None, numberOfFeatures=262144, oneHotEncodeCategoricals=True):
        super(Featurize, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.Featurize")
        self.allowImages = Param(self, "allowImages", "allowImages: Allow featurization of images (default: false)")
        self._setDefault(allowImages=False)
        self.featureColumns = Param(self, "featureColumns", "featureColumns: Feature columns")
        self.numberOfFeatures = Param(self, "numberOfFeatures", "numberOfFeatures: Number of features to hash string columns to (default: 262144)")
        self._setDefault(numberOfFeatures=262144)
        self.oneHotEncodeCategoricals = Param(self, "oneHotEncodeCategoricals", "oneHotEncodeCategoricals: One-hot encode categoricals (default: true)")
        self._setDefault(oneHotEncodeCategoricals=True)
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, allowImages=False, featureColumns=None, numberOfFeatures=262144, oneHotEncodeCategoricals=True):
        """
        Set the (keyword only) parameters

        Args:

            allowImages (bool): Allow featurization of images (default: false)
            featureColumns (dict): Feature columns
            numberOfFeatures (int): Number of features to hash string columns to (default: 262144)
            oneHotEncodeCategoricals (bool): One-hot encode categoricals (default: true)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setAllowImages(self, value):
        """

        Args:

            allowImages (bool): Allow featurization of images (default: false)

        """
        self._set(allowImages=value)
        return self


    def getAllowImages(self):
        """

        Returns:

            bool: Allow featurization of images (default: false)
        """
        return self.getOrDefault(self.allowImages)


    def setFeatureColumns(self, value):
        """

        Args:

            featureColumns (dict): Feature columns

        """
        self._set(featureColumns=value)
        return self


    def getFeatureColumns(self):
        """

        Returns:

            dict: Feature columns
        """
        return self.getOrDefault(self.featureColumns)


    def setNumberOfFeatures(self, value):
        """

        Args:

            numberOfFeatures (int): Number of features to hash string columns to (default: 262144)

        """
        self._set(numberOfFeatures=value)
        return self


    def getNumberOfFeatures(self):
        """

        Returns:

            int: Number of features to hash string columns to (default: 262144)
        """
        return self.getOrDefault(self.numberOfFeatures)


    def setOneHotEncodeCategoricals(self, value):
        """

        Args:

            oneHotEncodeCategoricals (bool): One-hot encode categoricals (default: true)

        """
        self._set(oneHotEncodeCategoricals=value)
        return self


    def getOneHotEncodeCategoricals(self):
        """

        Returns:

            bool: One-hot encode categoricals (default: true)
        """
        return self.getOrDefault(self.oneHotEncodeCategoricals)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.Featurize"

    @staticmethod
    def _from_java(java_stage):
        module_name=Featurize.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".Featurize"
        return from_java(java_stage, module_name)

    def _create_model(self, java_model):
        return PipelineModel(java_model)


class PipelineModel(ComplexParamsMixin, JavaModel, JavaMLWritable, JavaMLReadable):
    """
    Model fitted by :class:`Featurize`.

    This class is left empty on purpose.
    All necessary methods are exposed through inheritance.
    """

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "org.apache.spark.ml.PipelineModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=PipelineModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".PipelineModel"
        return from_java(java_stage, module_name)

